// exupery-resources/dist/implementation/generated/pareto/schemas/path/marshall.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Non_Normalized_Path = exports.Node_Path = exports.Context_Path = exports.Context_Subpath = exports.Start = exports.Up_Steps = void 0;
const _pa = __importStar(require("pareto-core-transformer"));
const Up_Steps = ($, $p) => ['text', ({
        'delimiter': ['backtick', null],
        'value': $p['value serializers']['default number']($, null),
    })];
exports.Up_Steps = Up_Steps;
const Start = ($, $p) => ['state', _pa.cc($, ($) => {
        switch ($[0]) {
            case 'absolute': return _pa.ss($, ($) => ({
                'state': "absolute",
                'value': ['nothing', null],
            }));
            case 'relative': return _pa.ss($, ($) => ({
                'state': "relative",
                'value': ['verbose group', _pa.dictionary_literal({
                        'up steps': _pa.cc($['up steps'], ($) => (0, exports.Up_Steps)($, {
                            'value serializers': $p['value serializers'],
                        })),
                    })],
            }));
            default: return _pa.au($[0]);
        }
    })];
exports.Start = Start;
const Context_Subpath = ($, $p) => ['list', $.map(($) => ['text', ({
            'delimiter': ['quote', null],
            'value': $,
        })])];
exports.Context_Subpath = Context_Subpath;
const Context_Path = ($, $p) => ['verbose group', _pa.dictionary_literal({
        'start': _pa.cc($['start'], ($) => (0, exports.Start)($, {
            'value serializers': $p['value serializers'],
        })),
        'subpath': _pa.cc($['subpath'], ($) => (0, exports.Context_Subpath)($, {
            'value serializers': $p['value serializers'],
        })),
    })];
exports.Context_Path = Context_Path;
const Node_Path = ($, $p) => ['verbose group', _pa.dictionary_literal({
        'context': _pa.cc($['context'], ($) => (0, exports.Context_Path)($, {
            'value serializers': $p['value serializers'],
        })),
        'node': _pa.cc($['node'], ($) => ['text', ({
                'delimiter': ['quote', null],
                'value': $,
            })]),
    })];
exports.Node_Path = Node_Path;
const Non_Normalized_Path = ($, $p) => ['verbose group', _pa.dictionary_literal({
        'leading slash': _pa.cc($['leading slash'], ($) => ['text', ({
                'delimiter': ['backtick', null],
                'value': $p['value serializers']['boolean']($, null),
            })]),
        'segments': _pa.cc($['segments'], ($) => ['list', $.map(($) => ['state', _pa.cc($, ($) => {
                    switch ($[0]) {
                        case 'parent': return _pa.ss($, ($) => ({
                            'state': "parent",
                            'value': ['nothing', null],
                        }));
                        case 'child': return _pa.ss($, ($) => ({
                            'state': "child",
                            'value': ['text', ({
                                    'delimiter': ['quote', null],
                                    'value': $,
                                })],
                        }));
                        case 'current': return _pa.ss($, ($) => ({
                            'state': "current",
                            'value': ['nothing', null],
                        }));
                        case 'nothing': return _pa.ss($, ($) => ({
                            'state': "nothing",
                            'value': ['nothing', null],
                        }));
                        default: return _pa.au($[0]);
                    }
                })])]),
        'trailing slash': _pa.cc($['trailing slash'], ($) => ['text', ({
                'delimiter': ['backtick', null],
                'value': $p['value serializers']['boolean']($, null),
            })]),
    })];
exports.Non_Normalized_Path = Non_Normalized_Path;
