// exupery-resources/dist/implementation/generated/pareto/schemas/path/serialize.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Non_Normalized_Path = exports.Node_Path = exports.Context_Path = exports.Context_Subpath = exports.Start = exports.Up_Steps = void 0;
const _i_serialize = __importStar(require("../../generic/serialize"));
const _i_marshall = __importStar(require("./marshall"));
const Up_Steps = ($, $p) => _i_serialize.Document(_i_marshall.Up_Steps($, {
    'value serializers': $p['value serializers'],
}));
exports.Up_Steps = Up_Steps;
const Start = ($, $p) => _i_serialize.Document(_i_marshall.Start($, {
    'value serializers': $p['value serializers'],
}));
exports.Start = Start;
const Context_Subpath = ($, $p) => _i_serialize.Document(_i_marshall.Context_Subpath($, {
    'value serializers': $p['value serializers'],
}));
exports.Context_Subpath = Context_Subpath;
const Context_Path = ($, $p) => _i_serialize.Document(_i_marshall.Context_Path($, {
    'value serializers': $p['value serializers'],
}));
exports.Context_Path = Context_Path;
const Node_Path = ($, $p) => _i_serialize.Document(_i_marshall.Node_Path($, {
    'value serializers': $p['value serializers'],
}));
exports.Node_Path = Node_Path;
const Non_Normalized_Path = ($, $p) => _i_serialize.Document(_i_marshall.Non_Normalized_Path($, {
    'value serializers': $p['value serializers'],
}));
exports.Non_Normalized_Path = Non_Normalized_Path;
