// exupery-resources/dist/implementation/generated/pareto/schemas/path/unmarshall.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Non_Normalized_Path = exports.Node_Path = exports.Context_Path = exports.Context_Subpath = exports.Start = exports.Up_Steps = void 0;
const _pa = __importStar(require("pareto-core-refiner"));
const _i_generic = __importStar(require("../../generic/unmarshall"));
const Up_Steps = ($, $p) => _i_generic.process_number($, {
    'deserializer': $p['value deserializers']['default number'],
});
exports.Up_Steps = Up_Steps;
const Start = ($, $p) => _i_generic.process_unconstrained_state_group($, {
    'states': _pa.dictionary_literal({
        'absolute': ($) => ['absolute', _i_generic.process_nothing($, null)],
        'relative': ($) => ['relative', _i_generic.process_group($, {
                'properties': ($) => ({
                    'up steps': _pa.cc(_i_generic.get_entry($, {
                        'key': "up steps",
                    }), ($) => (0, exports.Up_Steps)($, {
                        'value deserializers': $p['value deserializers'],
                    })),
                }),
            })],
    }),
});
exports.Start = Start;
const Context_Subpath = ($, $p) => _i_generic.process_unconstrained_list($, {
    'value': ($) => _i_generic.process_text($, null),
});
exports.Context_Subpath = Context_Subpath;
const Context_Path = ($, $p) => _i_generic.process_group($, {
    'properties': ($) => ({
        'start': _pa.cc(_i_generic.get_entry($, {
            'key': "start",
        }), ($) => (0, exports.Start)($, {
            'value deserializers': $p['value deserializers'],
        })),
        'subpath': _pa.cc(_i_generic.get_entry($, {
            'key': "subpath",
        }), ($) => (0, exports.Context_Subpath)($, {
            'value deserializers': $p['value deserializers'],
        })),
    }),
});
exports.Context_Path = Context_Path;
const Node_Path = ($, $p) => _i_generic.process_group($, {
    'properties': ($) => ({
        'context': _pa.cc(_i_generic.get_entry($, {
            'key': "context",
        }), ($) => (0, exports.Context_Path)($, {
            'value deserializers': $p['value deserializers'],
        })),
        'node': _pa.cc(_i_generic.get_entry($, {
            'key': "node",
        }), ($) => _i_generic.process_text($, null)),
    }),
});
exports.Node_Path = Node_Path;
const Non_Normalized_Path = ($, $p) => _i_generic.process_group($, {
    'properties': ($) => ({
        'leading slash': _pa.cc(_i_generic.get_entry($, {
            'key': "leading slash",
        }), ($) => _i_generic.process_boolean($, {
            'deserializer': $p['value deserializers']['boolean'],
        })),
        'segments': _pa.cc(_i_generic.get_entry($, {
            'key': "segments",
        }), ($) => _i_generic.process_unconstrained_list($, {
            'value': ($) => _i_generic.process_unconstrained_state_group($, {
                'states': _pa.dictionary_literal({
                    'parent': ($) => ['parent', _i_generic.process_nothing($, null)],
                    'child': ($) => ['child', _i_generic.process_text($, null)],
                    'current': ($) => ['current', _i_generic.process_nothing($, null)],
                    'nothing': ($) => ['nothing', _i_generic.process_nothing($, null)],
                }),
            }),
        })),
        'trailing slash': _pa.cc(_i_generic.get_entry($, {
            'key': "trailing slash",
        }), ($) => _i_generic.process_boolean($, {
            'deserializer': $p['value deserializers']['boolean'],
        })),
    }),
});
exports.Non_Normalized_Path = Non_Normalized_Path;
