// pareto-core-deserializer/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create_iterator = exports.text_to_character_list = exports.integer_division = exports.unreachable_code_path = exports.build_list = exports.build_text = exports.block = void 0;
/**
 * these functions coming from core-internals should be exposed for deserializer development
 */
var pareto_core_internals_1 = require("pareto-core-internals");
Object.defineProperty(exports, "block", { enumerable: true, get: function () { return pareto_core_internals_1.block; } });
Object.defineProperty(exports, "build_text", { enumerable: true, get: function () { return pareto_core_internals_1.build_text; } });
Object.defineProperty(exports, "build_list", { enumerable: true, get: function () { return pareto_core_internals_1.build_list; } });
Object.defineProperty(exports, "unreachable_code_path", { enumerable: true, get: function () { return pareto_core_internals_1.panic; } });
Object.defineProperty(exports, "integer_division", { enumerable: true, get: function () { return pareto_core_internals_1.integer_division; } });
Object.defineProperty(exports, "text_to_character_list", { enumerable: true, get: function () { return pareto_core_internals_1.text_to_character_list; } });
Object.defineProperty(exports, "create_iterator", { enumerable: true, get: function () { return pareto_core_internals_1.create_iterator; } });
