// pareto-core-transformer/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.group_list = exports.group_dictionary = exports.dictionary_literal = exports.list_literal = exports.unreachable_code_path = exports.deprecated_panic = exports.ss = exports.au = exports.cc = exports.not_set = exports.set = void 0;
/**
 * these functions coming from core-internals should be exposed for transformer development
 */
var pareto_core_internals_1 = require("pareto-core-internals");
Object.defineProperty(exports, "set", { enumerable: true, get: function () { return pareto_core_internals_1.set; } });
Object.defineProperty(exports, "not_set", { enumerable: true, get: function () { return pareto_core_internals_1.not_set; } });
Object.defineProperty(exports, "cc", { enumerable: true, get: function () { return pareto_core_internals_1.cc; } });
Object.defineProperty(exports, "au", { enumerable: true, get: function () { return pareto_core_internals_1.au; } });
Object.defineProperty(exports, "ss", { enumerable: true, get: function () { return pareto_core_internals_1.ss; } });
// block,
/**
 * panic should no longer be used,
 * if it is used in a transformation, this means it is not a transformation but a refinement
 */
Object.defineProperty(exports, "deprecated_panic", { enumerable: true, get: function () { return pareto_core_internals_1.panic; } });
Object.defineProperty(exports, "unreachable_code_path", { enumerable: true, get: function () { return pareto_core_internals_1.panic; } });
Object.defineProperty(exports, "list_literal", { enumerable: true, get: function () { return pareto_core_internals_1.list_literal; } });
Object.defineProperty(exports, "dictionary_literal", { enumerable: true, get: function () { return pareto_core_internals_1.dictionary_literal; } });
Object.defineProperty(exports, "group_dictionary", { enumerable: true, get: function () { return pareto_core_internals_1.group_dictionary; } });
Object.defineProperty(exports, "group_list", { enumerable: true, get: function () { return pareto_core_internals_1.group_list; } });
