// pareto-core-internals/dist/deprecated_refinement_result/create_failure_refinement_result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__deprecated_create_failure_refinement_result = __deprecated_create_failure_refinement_result;
class Failure_Refinement_Result_Class {
    constructor(error) {
        this.error = error;
    }
    transform(result_transformer, error_transformer) {
        return error_transformer(this.error);
    }
    transform_result(transformer) {
        return new Failure_Refinement_Result_Class(this.error);
    }
    deprecated_transform_error(error_transformer) {
        return new Failure_Refinement_Result_Class(error_transformer(this.error));
    }
    deprecated_refine_old_without_error_transformation(refiner) {
        return new Failure_Refinement_Result_Class(this.error);
    }
    deprecated_refine_old(stager, error_transformer) {
        return new Failure_Refinement_Result_Class(this.error);
    }
    rework_error_temp(error_reworker, rework_error_transformer) {
        return error_reworker(this.error).transform(($) => new Failure_Refinement_Result_Class($), ($) => new Failure_Refinement_Result_Class(rework_error_transformer($)));
    }
    __extract_data(on_result, on_error) {
        on_error(this.error);
    }
}
function __deprecated_create_failure_refinement_result(error) {
    return new Failure_Refinement_Result_Class(error);
}
