// pareto-core-internals/dist/deprecated_refinement_result/create_success_refinement_result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__deprecated_create_success_refinement_result = __deprecated_create_success_refinement_result;
class Success_Refinement_Result_Class {
    constructor(output) {
        this.output = output;
    }
    transform(result_transformer, error_transformer) {
        return result_transformer(this.output);
    }
    transform_result(transformer) {
        return new Success_Refinement_Result_Class(transformer(this.output));
    }
    deprecated_transform_error(error_transformer) {
        return new Success_Refinement_Result_Class(this.output);
    }
    deprecated_refine_old_without_error_transformation(refiner) {
        return refiner(this.output);
    }
    deprecated_refine_old(refiner, error_transformer) {
        //the error transform operation is here to satisfy the signature
        return refiner(this.output).deprecated_transform_error(error_transformer);
    }
    rework_error_temp(error_reworker, rework_error_transformer) {
        return new Success_Refinement_Result_Class(this.output);
    }
    __extract_data(on_result, on_error) {
        on_result(this.output);
    }
}
function __deprecated_create_success_refinement_result(output) {
    return new Success_Refinement_Result_Class(output);
}
