// pareto-core-internals/dist/dictionary_literal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dictionary_literal = dictionary_literal;
const set_1 = require("./set");
const not_set_1 = require("./not_set");
const list_literal_1 = require("./list_literal");
/**
 * this is an implementation, not public by design
 */
class Dictionary {
    constructor(source) {
        this.source = source;
    }
    map($v) {
        return new Dictionary(this.source.map(($) => {
            return {
                key: $.key,
                value: $v($.value, $.key)
            };
        }));
    }
    to_list(handle_value) {
        return (0, list_literal_1.list_literal)(this.source.map(($) => {
            return handle_value($.value, $.key);
        }));
    }
    filter(handle_value) {
        const out = [];
        this.source.forEach(($) => {
            const result = handle_value($.value, $.key);
            result.map((new_value) => {
                out.push({
                    key: $.key,
                    value: new_value
                });
            });
        });
        return new Dictionary(out);
    }
    get_entry(key) {
        for (let i = 0; i !== this.source.length; i += 1) {
            const element = this.source[i];
            if (element.key === key) {
                return (0, set_1.set)(element.value);
            }
        }
        return (0, not_set_1.not_set)();
    }
    get_number_of_entries() {
        return this.source.length;
    }
    is_empty() {
        return this.source.length === 0;
    }
}
/**
 * returns a Exupery dictionary
 *
 * why is this not the constructor? to call a constructor, you have to use the keyword 'new'. Exupery doesn't use the concept of a class so that keyword should be avoided
 * @param source An object literal
 * @returns
 */
function dictionary_literal(source) {
    //first we clone the source data so that changes to that source will have no impact on this implementation.
    //only works if the set does not become extremely large
    function create_dictionary_as_array(source) {
        const imp = [];
        Object.keys(source).forEach((key) => {
            imp.push({ key: key, value: source[key] });
        });
        return imp;
    }
    const daa = create_dictionary_as_array(source);
    return new Dictionary(daa);
}
