// pareto-core-internals/dist/list_literal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.list_literal = list_literal;
const not_set_1 = require("./not_set");
const set_1 = require("./set");
/**
 * this is an implementation, not public by design
 * If you feel the need to rename this class, don't rename it to 'Array',
 * it will break the 'instanceOf Array' test
 */
class List_Class {
    constructor(data) {
        this.data = data;
    }
    map($v) {
        return list_literal(this.data.map((entry) => {
            return $v(entry);
        }));
    }
    get_number_of_elements() {
        return this.data.length;
    }
    filter(handle_value) {
        const out = [];
        this.data.forEach(($) => {
            handle_value($).map(($) => {
                out.push($);
            });
        });
        return list_literal(out);
    }
    is_empty() {
        return this.data.length === 0;
    }
    append_element(new_element) {
        const new_data = this.data.slice();
        new_data.push(new_element);
        return new List_Class(new_data);
    }
    prepend_element(new_element) {
        const new_data = this.data.slice();
        new_data.unshift(new_element);
        return new List_Class(new_data);
    }
    reverse() {
        const new_data = this.data.slice();
        new_data.reverse();
        return new List_Class(new_data);
    }
    flatten(handle_element) {
        const out = [];
        this.data.forEach(($) => {
            const inner_list = handle_element($);
            inner_list.__for_each((inner_entry) => {
                out.push(inner_entry);
            });
        });
        return list_literal(out);
    }
    //internal methods
    __for_each($i) {
        this.data.forEach(($) => {
            $i($);
        });
    }
    __get_element_at(index) {
        if (index < 0 || index >= this.data.length) {
            return (0, not_set_1.not_set)();
        }
        return (0, set_1.set)(this.data[index]);
    }
    __get_raw_copy() {
        return this.data.slice();
    }
}
/**
 * returns a Exupery array

 * @param source An array literal
 * @returns
 */
function list_literal(source) {
    if (!(source instanceof Array)) {
        throw new Error("invalid input in 'list_literal'");
    }
    const data = source.slice(); //create a copy
    return new List_Class(data);
}
