// pareto-core-query/dist/create_query_result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__create_query_result = __create_query_result;
class Query_Result_Class {
    constructor(executer) {
        this.executer = executer;
        this.query_result = null;
    }
    transform_result(transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                on_result(transformer($));
            }, on_error);
        });
    }
    deprecated_transform_error(error_transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(on_result, ($) => {
                on_error(error_transformer($));
            });
        });
    }
    query_without_error_transformation(queryer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                queryer($).__extract_data(on_result, on_error);
            }, on_error);
        });
    }
    query(queryer, error_transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                queryer($).__extract_data(on_result, (queryer_error) => {
                    on_error(error_transformer(queryer_error));
                });
            }, on_error);
        });
    }
    deprecated_refine_old_without_error_transformation(refiner) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                refiner($).__extract_data(on_result, on_error);
            }, on_error);
        });
    }
    deprecated_refine_old(refiner, error_transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                refiner($).__extract_data(on_result, (stager_error) => {
                    on_error(error_transformer(stager_error));
                });
            }, on_error);
        });
    }
    rework_error_temp(error_reworker, rework_error_transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(on_result, ($) => {
                error_reworker($).__extract_data((new_target_error) => {
                    on_error(new_target_error);
                }, (rework_error) => {
                    on_error(rework_error_transformer(rework_error));
                });
            });
        });
    }
    __extract_data(on_result, on_error) {
        this.executer(on_result, on_error);
    }
}
function __create_query_result(executer) {
    return new Query_Result_Class(executer);
}
