// pareto-core-serializer/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.get_text_length = exports.ss = exports.au = exports.cc = exports.integer_division = exports.unreachable_code_path = exports.build_list = exports.build_text = void 0;
/**
 * these functions coming from core-internals should be exposed for serializer development
 */
var pareto_core_internals_1 = require("pareto-core-internals");
Object.defineProperty(exports, "build_text", { enumerable: true, get: function () { return pareto_core_internals_1.build_text; } });
Object.defineProperty(exports, "build_list", { enumerable: true, get: function () { return pareto_core_internals_1.build_list; } });
Object.defineProperty(exports, "unreachable_code_path", { enumerable: true, get: function () { return pareto_core_internals_1.panic; } });
Object.defineProperty(exports, "integer_division", { enumerable: true, get: function () { return pareto_core_internals_1.integer_division; } });
Object.defineProperty(exports, "cc", { enumerable: true, get: function () { return pareto_core_internals_1.cc; } });
Object.defineProperty(exports, "au", { enumerable: true, get: function () { return pareto_core_internals_1.au; } });
Object.defineProperty(exports, "ss", { enumerable: true, get: function () { return pareto_core_internals_1.ss; } });
const get_text_length = (text) => {
    return text.length;
};
exports.get_text_length = get_text_length;
