// exupery-resources/dist/implementation/generated/pareto/generic/resolve.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolve_ordered_dictionary = exports.resolve_dense_ordered_dictionary = exports.resolve_dense_dictionary = exports.resolve_dictionary = exports.resolve_path = exports.get_entry = exports.get_entry_from_stack = exports.push_stack = exports.get_possibly_circular_dependent_sibling_entry = exports.dictionary_to_lookup = exports.abort = void 0;
const _ea = __importStar(require("pareto-core-refiner"));
const _ps = __importStar(require("pareto-core-serializer"));
const abort = (location, type, location_to_string) => {
    return _ea.fixme_abort(_ea.cc(type, ($) => {
        switch ($[0]) {
            case 'no such entry': return _ea.ss($, ($) => `no such entry: '${$['key']}'`);
            case 'missing denseness entry': return _ea.ss($, ($) => `missing denseness entry: '${$['key']}'`);
            case 'circular dependency': return _ea.ss($, ($) => {
                const keys = _ps.build_text(($i) => {
                    $['keys'].__for_each(($) => {
                        $i['add snippet'](` '${$}', `);
                    });
                });
                return `circular dependency: (${keys})`;
            });
            case 'no context lookup': return _ea.ss($, ($) => `no context lookup`);
            case 'index out of bounds': return _ea.ss($, ($) => `index out of bounds, ${$['up steps taken']}`);
            case 'no element found at index': return _ea.ss($, ($) => `no element found at index, ${$['index']}`);
            default: return _ea.au($[0]);
        }
    }), ` @ ${location_to_string(location)}`);
};
exports.abort = abort;
const dictionary_to_lookup = ($, $p) => {
    return _ea.set($.map(($) => (['resolved', $])));
};
exports.dictionary_to_lookup = dictionary_to_lookup;
const get_possibly_circular_dependent_sibling_entry = ($, $p) => {
    return $.transform(($) => ({
        'key': $p.reference.key,
        'entry': $.get_entry($p.reference.key).transform(($) => $, () => (0, exports.abort)($p.reference.location, ['no such entry', { 'key': $p.reference.key }], $p['location 2 string']))
    }), () => (0, exports.abort)($p.reference.location, ['no context lookup', null], $p['location 2 string']));
};
exports.get_possibly_circular_dependent_sibling_entry = get_possibly_circular_dependent_sibling_entry;
const push_stack = ($, $p) => {
    return _ea.build_list(($i) => {
        $i['add list']($);
        $i['add element']($p['element']);
    });
};
exports.push_stack = push_stack;
const get_entry_from_stack = ($, $p) => {
    const ref = $p.reference;
    const get_entry_from_stack = (up_steps_taken) => {
        return $.__get_element_at($.get_number_of_elements() - 1 - up_steps_taken).transform(($) => {
            return $.transform(($) => {
                return $.get_entry(ref.key).transform(($) => _ea.cc($, ($) => {
                    switch ($[0]) {
                        case 'error': return _ea.ss($, ($) => get_entry_from_stack(up_steps_taken += 1));
                        case 'resolved': return _ea.ss($, ($) => ({
                            'key': ref.key,
                            'up steps': up_steps_taken,
                            'entry': $,
                        }));
                        default: return _ea.au($[0]);
                    }
                }), () => _ea.fixme_abort(`no clue yet of what is happening here`));
            }, () => (0, exports.abort)(ref.location, ['index out of bounds', { 'up steps taken': up_steps_taken }], $p['location 2 string']));
        }, () => (0, exports.abort)(ref.location, ['no element found at index', { 'index': up_steps_taken }], $p['location 2 string']));
    };
    return get_entry_from_stack(0);
};
exports.get_entry_from_stack = get_entry_from_stack;
const get_entry = ($, $p) => {
    return $.transform(($) => ({
        'key': $p.reference.key,
        'entry': $.get_entry($p.reference.key).transform(($) => _ea.cc($, ($) => {
            switch ($[0]) {
                case 'error': return _ea.ss($, ($) => _ea.cc($, ($) => {
                    switch ($[0]) {
                        case 'circular': return _ea.ss($, ($) => {
                            return (0, exports.abort)($p.reference.location, ['circular dependency', { 'keys': $ }], $p['location 2 string']);
                        });
                        default: return _ea.au($[0]);
                    }
                }));
                case 'resolved': return _ea.ss($, ($) => $);
                default: return _ea.au($[0]);
            }
        }), () => {
            return (0, exports.abort)($p.reference.location, ['no such entry', { 'key': $p.reference.key }], $p['location 2 string']);
        })
    }), () => (0, exports.abort)($p.reference.location, ['no context lookup', null], $p['location 2 string']));
};
exports.get_entry = get_entry;
const resolve_path = ($, $p) => {
    let current = {
        'list': _ea.list_literal([]),
        'result': {
            'data': $p.seed,
        },
    };
    $.list.__for_each(($) => {
        const result = $p.map($.element, current.result.data);
        const data = _ea.build_list(($i) => {
            current.list.__for_each(($) => {
                $i['add element']($);
            });
            $i['add element'](result.element);
        });
        current = {
            'list': data,
            'result': {
                'data': result.result,
            }
        };
    });
    return current;
};
exports.resolve_path = resolve_path;
const resolve_dictionary = ($, $p) => {
    return (0, exports.resolve_ordered_dictionary)($, $p).dictionary;
};
exports.resolve_dictionary = resolve_dictionary;
const resolve_dense_dictionary = ($, $p) => {
    return (0, exports.resolve_dense_ordered_dictionary)($, $p).dictionary;
};
exports.resolve_dense_dictionary = resolve_dense_dictionary;
const resolve_dense_ordered_dictionary = ($, $p) => {
    const location = $.location;
    const result = (0, exports.resolve_ordered_dictionary)($, $p);
    $p['denseness benchmark'].map(($) => {
        const validate_denseness = (benchmark, focus, location, location_to_string) => {
            benchmark.map(($, key) => {
                const benchmark = $;
                focus.get_entry(key).transform(($) => {
                }, () => {
                    (0, exports.abort)(location, ['missing denseness entry', { 'key': key }], $p['location 2 string']);
                });
            });
        };
        validate_denseness($p['denseness benchmark'], result.dictionary, location, $p['location 2 string']);
    });
    return result;
};
exports.resolve_dense_ordered_dictionary = resolve_dense_ordered_dictionary;
const resolve_ordered_dictionary = ($, $p) => {
    const dictionary_location = $.location;
    /**
     * this variable contains all the entries on which siblings have subscribed
     */
    const all_siblings_subscribed_entries = {};
    const finished = {};
    const ordered_list = _ea.build_list(($i) => {
        const source_dictionary = $;
        const status_dictionary = {};
        function process_entry($, location, key_of_entry_being_processed) {
            status_dictionary[key_of_entry_being_processed] = ['processing', null];
            const entry = $p.map({
                'key': key_of_entry_being_processed,
                'value': $,
                'location': location,
            }, {
                'possibly circular dependent siblings': _ea.set({
                    get_entry(key) {
                        //does the entry exist?
                        return source_dictionary.dictionary.get_entry(key).map(($) => {
                            //yes, it exists in the source dictionary
                            if (all_siblings_subscribed_entries[key] === undefined) {
                                all_siblings_subscribed_entries[key] = { 'entry': null };
                            }
                            const subscr = all_siblings_subscribed_entries[key];
                            return {
                                'get circular dependent': () => {
                                    if (subscr.entry === null) {
                                        return _ea.fixme_abort(`entry not set: ${key}`);
                                    }
                                    return subscr.entry;
                                }
                            };
                        });
                    },
                }),
                'not circular dependent siblings': _ea.set({
                    get_entry(key) {
                        const status = status_dictionary[key];
                        if (status === undefined) {
                            return source_dictionary.dictionary.get_entry(key).transform(($) => _ea.set(['resolved', process_entry($.entry, $.location, key)]), () => {
                                return _ea.not_set();
                                // throw new ResolveError("")
                            });
                        }
                        else {
                            const get_keys_of_entries_being_processed = () => {
                                return _ea.build_list(($i) => {
                                    _ea.dictionary_literal(status_dictionary).map(($, key) => {
                                        if ($[0] === 'processing') {
                                            $i['add element'](key);
                                        }
                                    });
                                });
                            };
                            return _ea.cc(status, (s) => {
                                switch (s[0]) {
                                    case 'failed':
                                        return _ea.ss(s, (s) => {
                                            //nothing to report
                                            return _ea.set(['error', ['circular', get_keys_of_entries_being_processed()]]);
                                            //return notSet()
                                        });
                                    case 'processing':
                                        if (key === key_of_entry_being_processed) {
                                            //$se.onError(`'${key}' is referencing itself`)
                                        }
                                        else {
                                            // const keys: string[] = []
                                            // Object.keys(status_dictionary).forEach((key) => {
                                            //     if (status_dictionary[key][0] === 'processing') {
                                            //         keys.push(key)
                                            //     }
                                            // })
                                            //$se.onError(`the following entries are referencing each other: ${keys.join(", ")}`)
                                        }
                                        status_dictionary[key_of_entry_being_processed] = ['failed', null];
                                        return _ea.set(['error', ['circular', get_keys_of_entries_being_processed()]]);
                                    case 'success':
                                        return _ea.set(['resolved', s[1]]);
                                    default: return _ea.au(s[0]);
                                }
                            });
                        }
                    },
                }),
            });
            finished[key_of_entry_being_processed] = entry;
            $i['add element']({
                'key': key_of_entry_being_processed,
                'value': entry,
            });
            status_dictionary[key_of_entry_being_processed] = ['success', entry];
            return entry;
        }
        $.dictionary.map(($, key) => {
            if (status_dictionary[key] === undefined) {
                process_entry($.entry, $.location, key);
            }
        });
        _ea.dictionary_literal(all_siblings_subscribed_entries).map(($, key) => {
            if (finished[key] === undefined) {
                _ea.fixme_abort(`implementation error: entry not resolved: ${key}`);
            }
            all_siblings_subscribed_entries[key].entry = finished[key];
        });
    });
    return {
        'dictionary': _ea.dictionary_literal(finished),
        'ordered list': ordered_list,
    };
};
exports.resolve_ordered_dictionary = resolve_ordered_dictionary;
